File: backend/routers/labels.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from services.person_annotator import CROPS_TOP_DIRNAME
from services.stage_resolver import final_stage_folder
from state import get_capture_manager, get_label_dataset, project_manager

router = APIRouter()


class DeleteImagesRequest(BaseModel):
    filenames: list[str]


class AssignLabelRequest(BaseModel):
    sample: str
    filenames: list[str]
    source: str = "crops"  # "crops" (full-body) atau "crops_top" (crop bagian atas)


@router.get("/api/projects/{project_id}/labels")
def list_labels(project_id: str):
    return get_label_dataset(project_id).list_labels()


@router.post("/api/projects/{project_id}/labels/{label}/assign")
def assign_label(project_id: str, label: str, req: AssignLabelRequest):
    if not label.strip():
        raise HTTPException(400, "Nama label tidak boleh kosong")
    if req.source not in ("crops", CROPS_TOP_DIRNAME):
        raise HTTPException(400, f"source harus 'crops' atau '{CROPS_TOP_DIRNAME}'")
    cm = get_capture_manager(project_id)
    if req.source == "crops":
        # "crops" sekarang berarti "ujung rantai tahap Crop saat ini" —
        # biasanya literal crops/ (checklist bawaan), tapi kalau project ini
        # sudah nambah tahap crop lebih dalam, ikut situ (lihat
        # stage_resolver.py) — bukan literal CROPS_DIRNAME lagi
        project = project_manager.get_project(project_id)
        stages = project["stages"] if project else []
        crops_dir = final_stage_folder(cm.label_dir(req.sample), stages)
    else:
        crops_dir = cm.label_dir(req.sample) / req.source
    if not crops_dir.exists():
        raise HTTPException(404, f"Sample '{req.sample}' belum punya crop di '{req.source}/'")
    assigned = get_label_dataset(project_id).assign(label, crops_dir, req.filenames)
    return {"assigned": assigned}


@router.get("/api/projects/{project_id}/labels/{label}/images")
def list_label_images(project_id: str, label: str, limit: int = 5000):
    return get_label_dataset(project_id).list_images(label, limit)


@router.delete("/api/projects/{project_id}/labels/{label}/images")
def delete_label_images(project_id: str, label: str, req: DeleteImagesRequest):
    deleted = get_label_dataset(project_id).delete_images(label, req.filenames)
    return {"deleted": deleted}


@router.delete("/api/projects/{project_id}/labels/{label}")
def delete_label(project_id: str, label: str):
    deleted = get_label_dataset(project_id).delete_label(label)
    if not deleted:
        raise HTTPException(404, f"Label '{label}' tidak ditemukan")
    return {"deleted": label}

File: frontend/src/api/labels.ts
import { p, request } from './client'
import type { LabelImage, LabelInfo } from './types'

// ---- Label (dataset klasifikasi hasil assign, per-project) ----

export const labelsApi = {
  listLabels: (projectId: string) => request<LabelInfo[]>(`${p(projectId)}/labels`),

  assignLabel: (
    projectId: string,
    label: string,
    sample: string,
    filenames: string[],
    source: 'crops' | 'crops_top' = 'crops',
  ) =>
    request<{ assigned: number }>(`${p(projectId)}/labels/${encodeURIComponent(label)}/assign`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ sample, filenames, source }),
    }),

  // hasil assign label (dataset/<label>/<filename>.jpg) — terpisah dari sample/
  labelImageUrl: (projectId: string, path: string) =>
    `/projects/${encodeURIComponent(projectId)}/dataset/${path}`,

  listLabelImages: (projectId: string, label: string, limit = 5000) =>
    request<LabelImage[]>(`${p(projectId)}/labels/${encodeURIComponent(label)}/images?limit=${limit}`),

  deleteLabelImages: (projectId: string, label: string, filenames: string[]) =>
    request<{ deleted: number }>(`${p(projectId)}/labels/${encodeURIComponent(label)}/images`, {
      method: 'DELETE',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ filenames }),
    }),

  deleteLabel: (projectId: string, label: string) =>
    request<{ deleted: string }>(`${p(projectId)}/labels/${encodeURIComponent(label)}`, {
      method: 'DELETE',
    }),
}
